Release 2.0.0

alessandropellegrini opened the pull request:
Realign `develop` branch to what is currently in `release-2.0.0`.    
This is the ultimate step to realign the development process to the internal branching model.

alessandropellegrini committed:
Merge branch 'hotfix-1.0.1'

alessandropellegrini committed:
Merge branch 'develop' into models

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge branch 'release-2.0.0' into models

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

readbeard committed:
Merge branch models into models

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
Realign ECS code to new list implementation

This commit fixes the way the list module is used by ECS and MPI
modules.

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Fix msgchannel size

A `msg_get()` call was decrementing the size of a message channel even
if the channel was empty.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

ael-code committed:
MPI: Get rid of MPI Datatype

The use of MPI datatype has been dropped. Messages between kernels are now exchanged as simple byte stream.
This allows to have the possibility to send variable length messages
while keepiping the code really simple.

The synchronus `MPI_Send` has been removed in favour of the `MPI_Get_Count`
approach

alessandropellegrini committed:
Merge branch 'models' into release-2.0.0

alessandropellegrini committed:
Fix the MPI Datatype used to exchange messages

This commit introduces a change to the MPI Datatype used to exchange
messages which reflects the changes done to the `msg_t` struct.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge pull request #78 from HPDCS/hotfix-mpi-datatype

Fix the MPI Datatype used to exchange messages

ael-code committed:
bugfix: list_insert_tail do not set prev & next to null

alessandropellegrini committed:
Merge pull request #80 from ael-code/fix_list_tail_insertion

bugfix: list_insert_tail do not set prev & next to null

alessandropellegrini committed:
Merge branch 'release-2.0.0' into models

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

readbeard committed:
Bugfix: zeroing wrong memory area

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
ECS code cleanup

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
ECS: fix red/white msg count

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Merge pull request #81 from ael-code/fix_bzero

Bugfix: zeroing wrong memory area

alessandropellegrini committed:
Merge branch 'release-2.0.0' into models

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

readbeard committed:
Fixed global variable issue

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Merge pull request #82 from ael-code/bytesend2

MPI: Get rid of MPI Datatype

readbeard committed:
Bugfix: LP bound overriding

The insertion of a message into the rendezvous queue requires a copy of the message itself otherwise we override the next/prev pointers.
Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
Bugfix: msg header allocation on stack :'(

Message header residing on stack was inserted into the output queue.

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Merge branch 'models' of github.com:HPDCS/ROOT-Sim into models

readbeard committed:
Bugfix: wrong ECS ifdef

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
Bugfix: wrong ECS ifdef

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Merge branch 'models' of github.com:HPDCS/ROOT-Sim into models

readbeard committed:
Bugfix: wrong thread initialization in ECS

When initializing threads with ECS enabled, a wrong number of lps was
passed, thus causing a wrong memory initialization in
IOCTL_SET_VM_RANGE.

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
Bugfix: wrong thread initialization in ECS

When initializing threads with ECS enabled, a wrong number of lps was
passed, thus causing a wrong memory initialization in
IOCTL_SET_VM_RANGE.

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Performance of slab memory management

This commit introduces a non-blocking retirement queue for buffers
obtained by the slab allocators, which allows to reduce concurrency
across threads when releasing memory.

This is still suboptimal. We should implement as well a backoff array in
case CAS'es start to fail too much.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Early release of msg buffers

With respect to the previous commit, if a message buffer was allocated
from a local slab, we early release it.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Slab memory release optimization

This commit introduces a different way to release message buffers from
slabs: the triber stack is detached with a single atomic exchange opertion,
and then all buffers (except the first, if present) are released.

We still need to make a thorough experimentation to determine whether
some variants of the proposed configurations might be combined together.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Transition from (M,N) to (M,1) msg channel

This commit makes the message channel data structure (M,1).

This change has a twofold effect:
* One of the two locks has been removed, with a slight performance
  improvement;
* A possible deadlock condition (although this never materialized) has
  been resolved, due to the fact that the lock acquisition was not
  respecting 2PL.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

ael-code committed:
fix MSG_META_SIZE

ael-code committed:
MPI: consume all messages

receive_remote_msgs() now tries to consume all pending messages instead
of just one.

ael-code committed:
Removed unused macro

ael-code committed:
Zeroing incoming message padding

alessandropellegrini committed:
Merge pull request #89 from ael-code/mpi_fixes

Inter-kernel message exchange fixes

Signed-off-by: Tommaso Tocci <tommaso.tocci@bsc.es>
Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge pull request #84 from alessandropellegrini/hotfix-performance-slab

Performance of slab memory management

alessandropellegrini committed:
Merge pull request #88 from alessandropellegrini/hotfix-msgchannel-1reader

Transition from (M,N) to (M,1) msg channel

ael-code committed:
Fixed tid type mismatch

alessandropellegrini committed:
Merge pull request #90 from ael-code/type_fix

Fixed tid type mismatch

readbeard committed:
Removed ECS prints to get better performance

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

readbeard committed:
Temp: new nosql model

This commit introduces a new version of the nosql model used to run test
using the distributed ECS protocol

Signed-off-by: Matteo Principe <matteo.principe92@gmail.com>

alessandropellegrini committed:
Merge branch 'models' of github.com:HPDCS/ROOT-Sim into models

alessandropellegrini committed:
Implemented tour pathfinder in topology library

This is based on a (memory-intensive!) A* implementation.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Code coverage

Add code coverage support by cover.io

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge branch 'codecov' into release-2.0.0

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Update README.md

Release 2.0.0 is now again distributed. The updates are reflected in the
current README.md file.

Commit:
mpi: changed iprobe to mprobe to increase performance.

Signed-off-by: Thananon Patinyasakdikul <tpatinya@utk.edu>

alessandropellegrini committed:
Merge branch 'develop' into release-2.0.0

alessandropellegrini committed:
Type safety in sequential simulation was crashing applications

The recent introduction of types for gids and lids was not taking into
account the fact that in sequential simulation they map to the same
values and the conversion arrays are thus not initialized.

This commit patches this bug and fixes #101

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Add Code of Conduct

This commit adds the Covenant Code of Conduct.

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Merge pull request #96 from thananon/mprobe

Changed iprobe to improbe

alessandropellegrini committed:
Merge branch 'master' into release-2.0.0

alessandropellegrini committed:
Don't bind sequential simulation on core 0

Sequential simulations do not need to obey core binding.
This fixes #20

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

readbeard commented on src/communication/mpi.c:
```diff
@@ -132,73 +114,46 @@ void send_remote_msg(msg_t *msg){
-	 * - `MPI_Recv` and `insert_bottom_half` need to be in the same critical section.
-	 *    messages need to be inserted in arrival order into the BH
-	 */
 	if(!spin_trylock(&msgs_lock))
```
```suggestion
	if(!spin_trylock(&msgs_lock))
```

i think we could remove this, we tried to run without the lock and everything was working

alessandropellegrini replied:
We tried as well, and it's actually crashing. We can remove the spinlock, but this depends on #102

readbeard commented on src/scheduler/scheduler.c:
```diff
@@ -506,7 +506,7 @@ void schedule(void) {
 #ifdef HAVE_CROSS_STATE
 	if(resume_execution && !is_blocked_state(LPS(lid)->state)) {
-		printf("ECS event is finished mark %llu !!!\n", LPS(lid)->wait_on_rendezvous);
+		printf("ECS event is finished at LP %d mark %llu !!!\n", lid_to_int(lid), LPS(lid)->wait_on_rendezvous);
```
```suggestion
		printf("ECS event is finished at LP %d mark %llu !!!\n", lid_to_int(lid), LPS(lid)->wait_on_rendezvous);
```
...removable

alessandropellegrini replied:
Done, although your other PR has commented out that line :smile:

readbeard commented on src/core/core.h:
```diff
@@ -180,14 +192,18 @@ typedef struct _msg_hdr_t {
 	int   			type;
 	unsigned long long	rendezvous_mark;	/// Unique identifier of the message, used for rendez-vous event
-	int			alloc_tid;
+	unsigned int		alloc_tid;
```
```suggestion
	unsigned int		alloc_tid;
```

Do we need to send it both in header and in actual msg?

alessandropellegrini replied:
We use this information to free a buffer of a positive message once it is annihilated, so it should be kept. The messages should be anyhow rearranged significantly. So far so good for me.

alessandropellegrini committed:
Remove a functional test in PCS which was generating a bug

We were violating the ordering of message generation explicitly to make
a functional test. This was inadverently pushed in the repo and
was reported in an issue. This commit undoes that change and fixes #87

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

readbeard commented on README.md:
```diff
@@ -84,9 +66,17 @@ executable, which is the model code already linked with the ROOT-sim library.
+
+To run in a distributed environment, you can use standard MPI commands,
+such as:    
+`mpiexec -n 2 -f hosts ./model --np 2 --nprc 16`
```
are we sure of `-f` ? The documentation says `--hostfile`

readbeard replied:
ps: that's a good case in which no "oversubscription" on nodes occurs...maybe it's clearer if we write also `--map-by node`?
But it's not really important.

alessandropellegrini replied:
Correct!

alessandropellegrini committed:
Merge pull request #93 from HPDCS/models

Slight changes to the current (non-final) version of ECS in release 2.0.0, and a new version of the nosql model.

alessandropellegrini committed:
Merge branch 'release-2.0.0' of github.com:HPDCS/ROOT-Sim into release-2.0.0

alessandropellegrini committed:
Fix 2.0.0 README

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>

alessandropellegrini committed:
Remove an unneeded parameter in an ECS print

Signed-off-by: Alessandro Pellegrini <pellegrini@dis.uniroma1.it>
